nsapass

```diff
@@ -401,19 +423,19 @@ try:
     elif args.command == 'get':
         nsa_get(args.name, args.index)
     elif args.command == 'add':
-        nsa_add(args.name, args.passwd, args.note)
+        locked(nsa_add)(args.name, args.passwd, args.note)
     elif args.command == 'gen':
-        nsa_gen(args.name, args.length, args.set, args.note)
+        locked(nsa_gen)(args.name, args.length, args.set, args.note)
     elif args.command == 'del':
-        nsa_del(args.name, args.index)
+        locked(nsa_del)(args.name, args.index)
     elif args.command == 'undel':
-        nsa_undel(args.name, args.index)
+        locked(nsa_undel)(args.name, args.index)
     elif args.command == 'mod':
-        nsa_mod(args.name, args.index, args.n, args.p, args.c)
+        locked(nsa_mod)(args.name, args.index, args.n, args.p, args.c)
     elif args.command == 'list':
         nsa_list(args.a)
     elif args.command == 'changepass':
-        nsa_changepass()
+        locked(nsa_changepass)()
```

Comment:
Note: `locked` would actually work as a decorator, but because some of the functions are used as building blocks for others (`nsa_add` in particular) I did the modification here in the command dispatch instead.

Reply:
ya.  i agree about the location, and your change's simplicity is attractive.

however, i have a question about a slightly different, approach.  just to get your opinion, no strong feelings towards/against it as your approach is already pretty neat. 

here is the alternative thought:

1. since file paths are constant, how about we define them in the configs?  no strong feelings.  just a suggestion/question.  e.g.:
```python
# location of encrypted passwords db, lock and temp files
DB = os.path.expanduser('~/.local/share/nsapass/db.enc')
DBLOCK = DB + '.lock'
DBTEMP = DB + '.temp'
```
2. we define a variable `is_locked = False` somewhere early on.  e.g.  after `KEY_NOTE ...`? (other places?)
3. within the dispatcher, we define `is_locked = True` before calling add, gen, del, undel, mod, changepass.
4. then modify `save` and `load` as follows:
```python
def load():
    if is_locked:
        print_debug('acquiring lock...')
        try:
            open(DBLOCK, 'x') # or 'xb'?  i don't know.
        except FileExistsError:
            print_err('lock file "{}" already exists'.format(DBLOCK))
            print_err('delete it if no nsapass instance is running') # simpler wording better?
            print_err('exiting...')
            sys.exit(1)
    print_info('loading db...')
    data = cmd(DECCMD)
    try:
        return json.loads(data)
    except json.JSONDecodeError:
        print_err('error decoding database "{}"'.format(DB))
        print_err('delete that file if you want to create a new database')
        print_err('exiting...')
        sys.exit(1)

def save(d):
    print_info('saving db...')
    data = json.dumps(d, indent=4)
    try:
        cmd([p.format(DBTEMP) for p in ENCCMD], stdin=data.encode())
        os.rename(DBTEMP, DB)
    finally:
        if is_locked:
            os.unlink(DBLOCK)
        try:
            os.unlink(DBTEMP)
        except FileNotFoundError:
            pass
```

this approach would have the advantage that `nsa_gen` can be execused normally even when, say, `nsa_add` is running, for the time period where `nsa_gen` is asking the user for whether he is happy with his password (which can be a long time period as it waits for input from the user).

a distadvantage that i see would be that the locking code is now distributed across mulitple places.  (you may see more disadvantages)

thoughts?

Reply:
1. no strong feelings either... I guess defining these as a constant helps avoid the ugly `{}` in `ENCCMD`. I'll change this too.
2. (and 3. and 4.) splitting the locking and the unlocking like this will fail to remove the lock file if the program terminates between load() and save(). the alternative I considered and decided against was to do the locking/unlocking as part of the command dispatch, because it would clutter that code, something like
```python
    if args.command is None:
        sys.stderr.write(banner)
        parser.print_help()
    elif args.command == 'get':
        nsa_get(args.name, args.index)
    elif args.command == 'list': #moved from below
        nsa_list(args.a)
    else:
        #take lock here
        try:
            if args.command == 'add:
                nsa_add(args.name, args.passwd, args.note)
            # elif...
        finally:
            #release lock here
```

Reply:
Aside, I guess one could use a global variable to make the lock reentrant, enabling the use of `locked` as a decorator. But I'm not sure whether it's worth it.

Reply:
one final thought:  since the dispatcher is already living in a `try:` block, how about we add `finally:` to it?  e.g. something like:

```python
try:
    if args.d:
        SHOW_DEBUG = True
    if args.z:
        SHOW_PASSWORDS = True
    if args.command is None:
        sys.stderr.write(banner)
        parser.print_help()
    elif args.command == 'get':
        nsa_get(args.username, args.index)
    elif args.command == 'add':
        is_locked = True
        nsa_add(args.username, args.passwd, args.note)
    elif args.command == 'gen':
        is_locked = True
        nsa_gen(args.username, args.l, args.s, args.c)
    elif args.command == 'del':
        is_locked = True
        nsa_del(args.username, args.index)
    elif args.command == 'undel':
        is_locked = True
        nsa_undel(args.username, args.index)
    elif args.command == 'mod':
        is_locked = True
        nsa_mod(args.username, args.index, args.u, args.p, args.c)
    elif args.command == 'list':
        nsa_list(args.a)
    elif args.command == 'changepass':
        is_locked = True
        nsa_changepass()
except KeyboardInterrupt:
    print_warn('received keyboard interrupt.')
    print_warn('exiting...')
finally:
    if is_locked:
        os.unlink(DBLOCK)
```
this way the code for `save(...)` won't change (only `load(...)`).

Reply:
Thanks for your comments btw! I'm learning things as well.